### ISSUE TITLE
Unable to self authorize CLI app

### ISSUE THREAD

**Issue by dan-da** [maidsafe/safe-api#322]
### Background
As an exercise, I decided to try and add --hexdump flag to 'safe cat'.

First though, I need a file to cat.  So following instructions in the README, I am trying to auth the CLI app so that I can *safe file add &lt;file>*.

### Issue
I am unable to auth the CLI app.  I have tried with and without --self-auth.    safe_vault and safe_authd are both running. When attempting without --self-auth I used *safe auth reqs* and *safe auth allow &lt;x>* in a separate terminal.

I have restarted both safe_authd (via safe auth start/stop) and safe_vault multiple times.  It is 100% repeatable.

Am I doing something wrong?

#### safe CLI
```
$ safe auth start
Starting SAFE Authenticator daemon (safe-authd)...
$ safe auth login --self-auth
Secret: 
Password: 
Sending login action request to authd...
Logged in successfully
Authorising CLI application...
[2019-11-15T21:16:43Z ERROR safe_api::api::quic_client] Connection lost: timed out
[2019-11-15T21:16:44Z ERROR safe_api::api::quic_client] Connection lost: timed out
Failed to self authorise: [Error] AuthdClientError - Failed to read response: connection closed: timed out
```

safe authd seems to be in a stuck state after this, because even 5 minutes later, I try:

```
$ safe auth status
Sending request to authd to obtain an status report...
[2019-11-15T21:30:30Z ERROR safe] safe-cli error: [Error] AuthdClientError - Failed to establish connection with authd: timed out
```

#### safe authd log
```
$ tail -f /tmp/safe-authd.out 
Initialising SAFE Authenticator services...
SAFE Authenticator services initialised sucessfully
Processing new incoming request: 'login'
Logging in to SAFE account...
Logged in successfully!
Processing new incoming request: 'subscribe'
Subscribing to authorisation requests notifications...
Subscription successful. Endpoint 'https://localhost:33002' will receive authorisation requests notifications (cert base path: Some("/home/danda/.config/safe-authd-client/net.maidsafe.cli"))
Processing new incoming request: 'authorise'
Authorising application...
The following application authorisation request was received:
AuthReq { app: AppExchangeInfo { id: "net.maidsafe.cli", scope: None, name: "SAFE CLI", vendor: "MaidSafe.net Ltd" }, app_container: false, app_permissions: AppPermissions { transfer_coins: true, perform_mutations: true, get_balance: true }, containers: {} }
Notifying subscriptor: https://localhost:33002
Processing new incoming request: 'allow'
Allowing authorisation request...
Authorisation request (742208197) allowed successfully
Let's request the authenticator lib to authorise the request '742208197'...
Subscriptor's response: Ok - auth req from app ID: net.maidsafe.cli ready to be notified to user
Decision obtained for auth req id: 742208197 - from app id: net.maidsafe.cli: None
```

#### safe_vault output
```
<startup omitted>
Waiting for connections ...
```

**Comment by bochaco** [comment]
@dan-da you are very likely hitting the issue on vault about a limitation on a single connection, do you see this error in your vault logs? : `Unable to deserialise challenge from 127.0.0.1:51829: invalid value: integer `5`, expected variant index 0 <= i < 3`, if so then you need the patch which is being worked on `vNext` branch of the safe_vault repo. 

To run that branch of vault you need this flags: `cargo run --features phase-one -- -f`

**Comment by bochaco** [comment]
I set some labels as this is a valid issue @dan-da , and it will be solved only whenever we get a new version of safe_vault released. So please let me know if it works with vNext branch and let's keep this open until we get new release of safe_vault.

**Comment by dan-da** [comment]
I don't see any errors/warnings in safe_vault console output.    Is there a vault logfile distinct from the stdout messages?  I see files under /tmp/safe_vault but nothing that appears to be a log.

Here is the full output:

```
$ ./safe_vault 
Checking target-arch... x86_64-unknown-linux-musl
Checking current version... v0.19.2
Checking latest released version... v0.19.2
Update status: `0.19.2`!
Our connection info:
{"peer_addr":"127.0.0.1:45338","peer_cert_der":[48,130,1,78,48,129,246,160,3,2,1,2,2,1,42,48,10,6,8,42,134,72,206,61,4,3,2,48,33,49,31,48,29,6,3,85,4,3,12,22,114,99,103,101,110,32,115,101,108,102,32,115,105,103,110,101,100,32,99,101,114,116,48,34,24,15,49,57,55,53,48,49,48,49,48,48,48,48,48,48,90,24,15,52,48,57,54,48,49,48,49,48,48,48,48,48,48,90,48,33,49,31,48,29,6,3,85,4,3,12,22,114,99,103,101,110,32,115,101,108,102,32,115,105,103,110,101,100,32,99,101,114,116,48,89,48,19,6,7,42,134,72,206,61,2,1,6,8,42,134,72,206,61,3,1,7,3,66,0,4,174,23,125,180,121,190,142,239,196,22,68,233,29,34,128,226,236,112,184,182,132,61,38,239,21,9,213,92,181,103,45,183,214,98,88,198,21,92,57,137,230,30,16,131,42,28,217,224,151,148,74,50,126,122,229,129,91,142,241,8,69,27,76,202,163,27,48,25,48,23,6,3,85,29,17,4,16,48,14,130,12,77,97,105,100,83,65,70,69,46,110,101,116,48,10,6,8,42,134,72,206,61,4,3,2,3,71,0,48,68,2,32,116,180,205,211,232,164,249,173,237,191,94,201,161,65,173,110,70,60,117,80,60,36,86,30,150,254,191,45,73,13,233,221,2,32,3,113,170,203,197,168,80,253,179,162,99,52,167,200,133,214,58,18,66,128,103,69,214,136,121,201,56,177,135,231,160,221]}

Writing connection info to: /home/user/.config/safe_vault/vault_connection_info.config
Waiting for connections ...
```

I have been running safe_vault that is latest release on github.  I will try building it and using the branch you mention...    thx.

**Comment by bochaco** [comment]
You need to run the vault with `RUST_LOG=trace` (or debug) to see the logs in stdout

**Comment by dan-da** [comment]
thx.  ok, now I see deserialization errors like:

```
INFO 2019-11-15T22:27:01Z [src/client_handler.rs:195] 0736da..: Unable to deserialise message from Client(Bls(8d13a0..)): tag for enum is not valid, found 246
```

**Comment by dan-da** [comment]
Update --  I got the vNext branch safe_vault built, restarted with the command you provided, and restarted authd, but login still fails.

```
$ safe auth login --self-auth
Secret: 
Password: 
Sending login action request to authd...
[2019-11-15T23:10:59Z ERROR safe] safe-cli error: [Error] AuthdError - SAFE Authenticator: [Error] AuthError - Failed to log in: SndError(NoSuchLoginPacket)
```

from safe_authd log
```
Processing new incoming request: 'login'
Logging in to SAFE account...
Error occurred when trying to log in: [Error] AuthError - Failed to log in: SndError(NoSuchLoginPacket)
```


from safe_vault output:

The log is quite large, but this is the only thing that seems an error.

```
DEBUG 2019-11-15T23:10:58Z [/home/danda/.cargo/git/checkouts/quic-p2p-9a2715614ed5c99d/db9f47c/src/utils.rs:93] ERROR in communication with peer 127.0.0.1:39073: Connection(ApplicationClosed { reason: ApplicationClose { error_code: 0, reason: b"" } }) - Connection Error: closed by peer: 0. Details: Incoming streams failed
INFO 2019-11-15T23:10:58Z [src/client_handler.rs:115] 93c50a..: Disconnected from Client(Bls(98e253..)) on 127.0.0.1:39073
```

I did have to change from rustc 1.37 to 1.38 in order to build safe_vault.   So safe-api was built with 1.37.  I doubt that's the issue, but mention it just in case.

**Comment by bochaco** [comment]
That error is because the account doesn't exist in the vault, please make sure you create one with `safe auth create-acc` then login and sefl-auth with it.

If you pull latest commit from safe-api you'll get some recent enhacements for those type of error messages

**Comment by dan-da** [comment]
hooray!    It worked.   thanks for your help @bochaco.  I will update/rebuild tonight.

and my little --hexdump hack even seems to work, first try.   :-)

```
$ safe cat --hexdump "safe://hbhyrydxhpah66r4a595drmr5kfxe1d8frp46g3qocg9gswg56ncr4e4im" | head -n 10
|7c446f63 756d656e 74617469 6f6e7c4c| |Documentation|L 00000000
|696e7578 2f6d6163 4f532f57 696e646f| inux/macOS/Windo 00000010
|77737c0a 7c3a2d2d 2d2d2d2d 2d2d2d2d| ws|.|:---------- 00000020
|2d3a7c3a 2d2d2d2d 2d2d2d2d 2d2d2d2d| -:|:------------ 00000030
|2d2d2d2d 2d3a7c0a 7c205b21 5b446f63| -----:|.| [![Doc 00000040
|756d656e 74617469 6f6e5d28 68747470| umentation](http 00000050
|733a2f2f 646f6373 2e72732f 73616665| s://docs.rs/safe 00000060
|2d636c69 2f626164 67652e73 7667295d| -cli/badge.svg)] 00000070
|28687474 70733a2f 2f646f63 732e7273| (https://docs.rs 00000080
|2f736166 652d636c 6929207c 205b215b| /safe-cli) | [![ 00000090
```

**Comment by bochaco** [comment]
ohh nice!! that's looking really cool man! keep it up, and send the PR for it when ready, I'll review it

**Comment by dan-da** [comment]
hmm, somehow 'safe' app became unauthorized again today.  is there a timeout?

Now the 'safe auth' is failing again.    safe_vault hasn't changed and is running the same code from vNext branch.   I did update safe-api to latest code from master though.  Maybe something in there broke it?

**Comment by bochaco** [comment]
This is expected to be fixed with https://github.com/maidsafe/safe_vault/pull/897, we can close once that's release in a new patch version of the vault

**Comment by bochaco** [comment]
Fixes with vault v0.20.0
